File: SecureStore.POS/Application/Services/StockService.cs
using SecureStore.POS.Application.DTOs;
using SecureStore.POS.Application.Interfaces;
using SecureStore.POS.Domain.Entities;
using SecureStore.POS.Domain.Enums;
using SecureStore.POS.Infrastructure.UnitOfWork;

namespace SecureStore.POS.Application.Services;

public class StockService : IStockService
{
    private readonly IUnitOfWork _uow;
    private readonly IAuditService _audit;

    public StockService(IUnitOfWork uow, IAuditService audit)
    {
        _uow = uow;
        _audit = audit;
    }

    public async Task<(bool success, string message)> AdjustStockAsync(StockAdjustmentDto dto)
    {
        if (string.IsNullOrWhiteSpace(dto.Reason))
            return (false, "A reason is required for stock adjustment.");
        var reason = dto.Reason.Trim();
        if (reason.Length > 500)
            return (false, "Stock adjustment reason must be 500 characters or fewer.");

        await using var transaction = await _uow.BeginTransactionAsync();
        try
        {
            var product = await _uow.Products.GetByIdAsync(dto.ProductID);
            if (product is null) return (false, "Product not found.");

            int oldStock = product.StockQuantity;
            int newStock = oldStock + dto.QuantityChange;

            if (newStock < 0)
                return (false, $"Adjustment would result in negative stock ({newStock}). Not allowed.");

            product.StockQuantity = newStock;
            product.UpdatedAt = DateTime.UtcNow;
            _uow.Products.Update(product);

            await _uow.Inventory.AddAsync(new InventoryTransaction
            {
                ProductID = product.ProductID,
                TransactionType = TransactionType.Adjustment,
                QuantityChange = dto.QuantityChange,
                OldStock = oldStock,
                NewStock = newStock,
                UserID = dto.UserID,
                Reason = reason,
                CreatedAt = DateTime.UtcNow
            });

            await _uow.SaveChangesAsync();
            await transaction.CommitAsync();

            await _audit.LogAsync("STOCK_ADJUSTED", "Product", dto.ProductID.ToString(),
                $"Stock for '{product.ProductName}' adjusted by {dto.QuantityChange:+#;-#;0}. " +
                $"Old: {oldStock}, New: {newStock}. Reason: {reason}", dto.UserID);

            return (true, $"Stock adjusted successfully. New stock: {newStock}");
        }
        catch (Exception ex)
        {
            try { await transaction.RollbackAsync(); } catch { }
            AppLogger.LogError(ex, "StockService.AdjustStockAsync");
            return (false, $"Stock adjustment failed: {ex.Message}");
        }
    }

    public async Task<IEnumerable<ProductDto>> GetLowStockProductsAsync()
    {
        var products = await _uow.Products.GetLowStockAsync();
        return products.Select(p => new ProductDto(
            p.ProductID, p.ProductName, p.Barcode, p.CategoryID,
            p.Category?.CategoryName ?? "", p.Brand,
            p.PurchasePrice, p.SellingPrice, p.StockQuantity,
            p.ReorderLevel, p.IsActive, p.IsLowStock,
            p.CreatedAt, p.UpdatedAt));
    }

    public async Task<IEnumerable<InventoryTransaction>> GetInventoryHistoryAsync(int productId) =>
        await _uow.Inventory.GetByProductAsync(productId);
}

File: pos-app/electron/services/stockService.ts
import { all, withTx, type Db } from '../database/database'
import { auditService } from './auditService'
import { backupService } from './backupService'
import { logger } from '../utils/logger'
import { PublicError, publicErrorMessage } from '../utils/safeErrors'
import type { StockAdjustmentRequest, ServiceResult } from '../../shared/types'

export const stockService = {
  adjustStock: async (req: StockAdjustmentRequest): Promise<ServiceResult> => {
    try {
      await withTx(async (tx: Db) => {
        if (!Number.isInteger(req.productId) || req.productId <= 0) throw new PublicError('Choose a valid product.')
        if (!Number.isInteger(req.quantityChange) || req.quantityChange === 0 || Math.abs(req.quantityChange) > 100000) {
          throw new PublicError('Enter a valid non-zero stock adjustment.')
        }
        const reason = String(req.reason ?? '').trim()
        if (reason.length < 3 || reason.length > 500) throw new PublicError('Reason must be between 3 and 500 characters.')

        const prod = await tx.get('SELECT StockQuantity, ProductName FROM Products WHERE ProductID = $1 AND IsActive = true', [req.productId]) as any
        if (!prod) throw new PublicError('Choose an active product.')

        const newStock = prod.StockQuantity + req.quantityChange
        if (newStock < 0) {
          throw new PublicError(`Adjustment would result in negative stock. Current: ${prod.StockQuantity}`)
        }

        await tx.run('UPDATE Products SET StockQuantity = $1, UpdatedAt = now() WHERE ProductID = $2', [newStock, req.productId])

        await tx.run(
          `INSERT INTO InventoryTransactions (ProductID, TransactionType, QuantityChange, OldStock, NewStock, UserID, Reason)
           VALUES ($1, 'Adjustment', $2, $3, $4, $5, $6)`,
          [req.productId, req.quantityChange, prod.StockQuantity, newStock, req.userId, reason]
        )

        const updatedProduct = await tx.get(
          'SELECT ProductID as productId, ProductName, Barcode, CategoryID as categoryId, Brand, PurchasePrice as purchasePrice, SellingPrice as sellingPrice, StockQuantity as stockQuantity, ReorderLevel as reorderLevel, IsActive, CreatedAt, UpdatedAt FROM Products WHERE ProductID = $1',
          [req.productId]
        )
        const backupResult = await backupService.insertProductBackup(updatedProduct as any, 'ADJUSTMENT', req.userId, { before: prod, after: updatedProduct }, tx)
        if (!backupResult.success) throw new PublicError(backupResult.message || 'Product backup failed.')

        await auditService.log('STOCK_ADJUSTED', 'Product', `Stock adjusted for ${prod.ProductName} by ${req.quantityChange}. Reason: ${reason}`, req.userId, req.productId.toString(), tx)
      })
      return { success: true, message: 'Stock adjusted successfully' }
    } catch (error: any) {
      logger.error('Stock adjustment failed', error)
      return { success: false, message: publicErrorMessage(error) }
    }
  },

  getHistory: async (productId: number) => {
    try {
      const data = await all(`
        SELECT t.*, u.Username as userName, p.ProductName as productName
        FROM InventoryTransactions t
        JOIN Users u ON t.UserID = u.UserID
        JOIN Products p ON t.ProductID = p.ProductID
        WHERE t.ProductID = $1
        ORDER BY t.CreatedAt DESC
        LIMIT 100
      `, [productId])
      return { success: true, data, message: '' }
    } catch (error: any) {
      return { success: false, message: publicErrorMessage(error), data: [] }
    }
  }
}
